### PR TITLE
#CP-32 create roles & set permissions

### PR DESCRIPTION
**Pull request by Yvad60**
### What the PR does
allows the admin to create the role and set its permissions, so that different users can only access & perform data/operations to which they are authorized.

### Description of Tasks to be completed

- [x] The Administrator should be able to create, update, and delete roles.
- [x] The administrator should be able to set and update the role's permissions.
- [x] Appropriate error messages are displayed on unsuccessful attempts
- [x] Skeleton UIs/screen are/is shown when the page loads/is loading.
- [ ] Created pages are responsive.

### How this should be manually tested
Clone the project repository linked [here](https://github.com/atlp-rwanda/phantom-fe-codebandits)

```
cd phantom-fe-codebandits/
git checkout ft-create-roles-set-permissions-cp-32
git pull 
npm install
npm run start
run http://localhost:3000 in your browser
```

### Relevant Trello Card
[#CP-32](https://trello.com/c/3bd3Q7a2/32-create-roles-set-permissions)

## Introduced changes
This PR introduce changes to the files not related to the tasks. The core behaviour of the files was not affected but it worth taking into considerations of the following:
- #### Aliases and path naming:
In this PR a handful number of `aliases` have been added to solve the issue of linking to files deep in modules
Those aliases are added into three core project files and one `vscode` workspace settings (more on that later).
#### Usage:
To define a custom alias that correspond to the given directory. Follow the naming convention based on the file. 

> Only code related to changes is shown here for **brevity**

- [x]  **`webpack.config.js`**

    ```
  resolve: {
    alias: {
      '@': path.resolve(__dirname, 'src'),
      '@pages': path.resolve(__dirname, 'src/pages'),
      '@components': path.resolve(__dirname, 'src/components'),
      '@utils': path.resolve(__dirname, 'src/utils'),
      '@reducers': path.resolve(__dirname, 'src/redux/reducers'),
      '@forms': path.resolve(__dirname, 'src/forms')
    },
    modules: ['node_modules'],
  },
   ```
- [x] **`.babelrc`**

```
 "plugins": [
        ["@babel/transform-runtime",
            {
                "module-resolver": {
                    "root": ["./src"],
                    "aliases": {
                        "@": "./src",
                        "@pages": "./src/pages",
                        "@components": "./src/components",
                        "@reducers": "./src/redux/reducers",
                        "@forms": "./src/forms",
                        "@utils": "./src/utils"
                    }
                }
            }
        ]
    ]
```

- [x] **`package.json`**

> Entry changed are related to jest configurations and can be configured inside jest own file like `jest.config.json`

```
    "moduleNameMapper": {
      "\\.(css|scss)$": "identity-obj-proxy",
      "^@/(.*)$": "<rootDir>/src/$1",
      "^@pages/(.*)$": "<rootDir>/src/pages/$1",
      "^@components/(.*)$": "<rootDir>/src/components/$1",
      "^@reducers/(.*)$": "<rootDir>/src/redux/reducers/$1",
      "^@forms/(.*)$": "<rootDir>/src/forms/$1",
      "^@utils/(.*)$": "<rootDir>/src/utils/$1",
      "^react(.*)$": "<rootDir>/node_modules/react$1"
    },
```
- #### Working with VS code `intellisence`
For VS Code to be aware of the imports under the aliases. An extension was added to the recommended one. If you get prompted for the extension install it or better [get it here](https://marketplace.visualstudio.com/items?itemName=ionutvmi.path-autocomplete). 

There is a `.vscode` folder added to share the settings for the workspace only. Note: These changes only affect this project, not your core user settings. If you prefer to change the whole settings. The additional lines are added or replaced when already exists:

```
// .vscode/settings.json  // prefer user settings>> ctrl+shift+P then type settings.json and choose user settings
{
    "javascript.suggest.paths": false,
    "path-autocomplete.extensionOnImport": true,
    "path-autocomplete.pathMappings": {
        "@": "${folder}/src",
        "@utils": "${folder}/src/utils",
        "@components": "${folder}/src/components",
        "@reducers": "${folder}/src/redux/reducers",
        "@forms": "${folder}/src/forms",
        "@pages": "${folder}/src/pages"
    },
}
```
### Conflict resolution
Take into consideration the above mentioned aliases.

@Yvad60 @samuel-shyaka-dus @Abdulkeza @feyton

### CODE REVIEW

**Review by samyshyaka — CHANGES_REQUESTED** [review]
This is great work. Find a better way to make the table responsive. For  example for smaller screens, you can display only a few columns per page then create a column title, name it **more** and make it a dropdown. once a user clicks on **more** they can choose a title form the dropdown (for example: Assigned bus) and the column will contain the data that correspond to the title.

**Review by Abdulkeza — DISMISSED** [review]
Great work! and I've seen that the issues that we saw were all solved.

**Review by samyshyaka — DISMISSED** [review]

File: src/Modals/ChangeRole.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    { name: 'Driver', info: 'Can start and stop car', value: 3 }
+  ];
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    const value = e.target.permission.value;
+    toast('Successfully change the user role to ');
+    const role = roles.filter((entry) => {
+      return entry.value == value;
+    });
+    console.log(role);
```
Comment (Jkadhuwa) [inline src/Modals/ChangeRole.js]: remove consoles.

File: src/pages/dashboard/RolesPage.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      <h1 className="text-2xl font-bold text-center justify-center font-rale">
+        Manage roles and permissions
+      </h1>
+      <div className="mt-6 p-5 rounded-md shadow-main">
+        <div className="w-full border rounded-sm p-2 ">
+          <div>
+            <ul className="grid grid-cols-[30%_60%] gap-1 border-b-2 font-raleway font-bold py-1 px-2">
+              <li>Role</li>
+              <ul className="grid grid-cols-2">
+                <li>Object</li>
```
Comment (Jkadhuwa) [inline src/pages/dashboard/RolesPage.js]: Think of  a better name here

File: src/Modals/ChangeRole.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          </option>
+          {roles.map((role) => {
+            return (
+              <option value={role.value} key={role.name}>
+                {role.name}
+              </option>
+            );
+          })}
+        </select>
+        <PrimaryButton type={'submit'} name={'Continue'} />
```
Comment (Jkadhuwa) [inline src/Modals/ChangeRole.js:76]: After changing role and clicking on the "Continue" button nothing happens after the toast

File: src/Modals/ChangeRole.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    setNewRole(role[0]);
+  };
+  return (
+    <div className="font-raleway">
+      <h1 className="font-bold font-raleway text-center">Change user's role</h1>
+      <hr />
+      <div>
+        <ul>
+          <li>
+            <b>Name</b>: {name}
```
Comment (Jkadhuwa) [inline src/Modals/ChangeRole.js:33]: Increase the size of the card, its so congested.

**Review by Jkadhuwa — CHANGES_REQUESTED** [review]
Nice work @feyton and @Yvad60 , left some comments below. Please make sure you remove all unnecessary console logs and fix your snapshot test.

File: src/pages/dashboard/RolesPage.js (hunk 2 above)
Comment (feyton) [reply src/pages/dashboard/RolesPage.js]: I guess using Model would be better 
What do you think

**Review by samyshyaka — APPROVED** [review]

**Review by Abdulkeza — APPROVED** [review]
Great job Fab!!

**Review by Jkadhuwa — APPROVED** [review]

**Pull request merged**